xclib/evaluation/xc_metrics.py

```diff
@@ -650,3 +649,139 @@ def __init__(self, true_labels, inv_propensity_scores=None,
             "Metrices() is deprecated; use Metrics().",
             category=FutureWarning)
         super().__init__(true_labels, inv_propensity_scores, remove_invalid)
+
+@nb.njit(parallel=True)
+def _fast_recall_with_indices(pred_indices, true_indices, true_indptr, k):
+    m = pred_indices.shape[0]
+    cum_intersections = np.zeros((m, k), dtype=np.float64)
+    gt_cardinality = np.zeros((m, 1), dtype= np.float64)
+    for i in nb.prange(pred_indices.shape[0]):
+        intersection = in1d(pred_indices[i][:k], np.unique(true_indices[true_indptr[i]: true_indptr[i + 1]]))
+        cum_intersections[i] = np.cumsum(intersection)
+        gt_cardinality[i, 0] = true_indptr[i + 1] - true_indptr[i]
+
+    recall = mean_rows(np.multiply(cum_intersections, 1 / gt_cardinality))
+    micro_recall = np.sum(cum_intersections, axis = 0) / np.sum(gt_cardinality)
+    return recall, micro_recall
+
+def fast_recall_with_indices(X_indices, true_labels, k):
+    """
+    Compute recall@k, microR@k for values 1...k, faster than using `recall`
+    Arguments:
+    ----------
+    X_indices: np.ndarray
+        2D numpy array with indices sorted in descending order according to score for each row
+    true_labels: csr_matrix
+        ground truth in sparse format
+    k: int
+        compute psprecision for ints in [1, k]
+    Returns:
+    -------
+    np.ndarray: recall values
+    """
+    return _fast_recall_with_indices(X_indices, true_labels.indices.astype(np.int64), true_labels.indptr, k)
+
+@nb.njit(parallel=True)
+def restict_preds_for_gt_calc(pred_indices, num_gt, k):
+    """
+    Returns top min(GT, k) indices for each data point
+    Arguments:
+    ----------
+    X_indices: np.ndarray
+        2D numpy array with indices sorted in descending order according to score for each row
+    num_gt: np.array
+        number of ground truth for each element
+    k: int
+        max number of preds to consider
+    Returns:
+    -------
+    np.ndarray:  top min(GT, k) indices for each data point
+    """
+    num_docs = pred_indices.shape[0]
+    high_lim = 100000000
```

Comment:
Can we please ask the user for the pad value instead of hardcoding it?

Reply:
pad value index is now taken as an input from the user (as a command line argument): (refer [here](https://github.com/kunaldahiya/pyxclib/blob/fc3bb71c57283ea28ef8f9377572534102a10918/xclib/examples/fast_evaluate.py#L56)) and then passed as an argument to this block of code.